cpp/src/tsne/distances.h

```diff
@@ -78,29 +79,29 @@ void normalize_distances(const int n, float *distances, const int n_neighbors,
  * @input param indices: The input sorted indices from KNN.
  * @input param n: The number of rows in the data X.
  * @input param k: The number of nearest neighbors you want.
- * @input param P_sum: The sum of P.
  * @input param exaggeration: How much early pressure you want the clusters in TSNE to spread out more.
- * @output param COO_Matrix: The final P + P.T output COO matrix.
+ * @output param VAL: Values of COO Matrix
+ * @output param COL: Column indices of COO Matrix
+ * @output param ROW: Row indices of COO Matrix
+ * @input param row_sizes: Size (2*n) array for CSR style Row pointers. Used in symmetrization
  * @input param stream: The GPU stream.
  * @input param handle: The GPU handle.
  */
 template <int TPB_X = 32>
 void symmetrize_perplexity(float *P, long *indices, const int n, const int k,
-                           const float P_sum, const float exaggeration,
+                           const float exaggeration,
                            MLCommon::Sparse::COO<float> *COO_Matrix,
+                           int *row_sizes,
                            cudaStream_t stream, const cumlHandle &handle) {
   // Perform (P + P.T) / P_sum * early_exaggeration
-  const float div = exaggeration / (2.0f * P_sum);
+  const float div = exaggeration / (2.0f * n);
```

Comment:
This change from `P_sum` to `n` looks like an actual algorithmic change. Could you explain it please?

Reply:
Oh so P_sum ie the sum(P) should be == n. Though due to roundoff error, this might not be the case (ie small epsilon differences). However, for efficiency reasons, atomicAdd(P_sum, P[i]) can be commented out if we just assume sum(P) = n.
Notice if sum(P) = n, then sum(P + P.T) = 2*sum(P) = 2*n.

Reply:
Thanks for the responses @danielhanchen

Reply:
No problems @zbjornson !